_posts/2017-04-30-React-Native-for-iOS-devs.markdown

```diff
@@ -0,0 +1,658 @@
+
+We choose TypeScript because of how well VS Code and TypeScript integrate. It provides an Xcode-like level of integration. Which, IMO, is a high baseline for expectations. It's definitely less polished, but it's Open Source - which is a very fair trade-off to me.
+
+The node community is great at automation, we have code formatters and language linters that will auto-fix your code as you press save. We have pre-commit, pre-push hooks that are set up automatically when you start using the project. It means you spend less time thinking about trivial details that add up. It's wonderful.
+
+Debugging our code is still a really tricky problem, we initially had all our debugging set up inside VS Code but eventually it got reliable and people moved back to `console.log` or using the debugger inside Chrome. For what it's worth, the type of code we've been writing has been much easier to debug as it is significantly simpler. It's not been a pressing problem.
+
+# Testing
+
+Testing was never a priority in the iOS world. I feel like it's always getting better with each Xcode version, particularly the change in Xcode 9 which allows tests to run in another simulator without turning off your app.
+
+We really put a lot of time and patience into our native testing on iOS. Coming from that world, to the absolute ease of testing in the JavaScript world is pretty breath-taking. Check out my [coverage of Jest's features][artsy-jest].
+
+There are two ways to write tests for your React Native code: in process and out of process. E.g. from the JavaScript side, or in native side.
+
+-   JS side: Choice of many Open Source test runners, built with hundreds of contributors involved in multi-year test runners.
+
+-   Native side: Probably one or two people making XCTest, one person trying to get some improvements in Xcode each year. [Semi-closed source][xctest].
+
+We tried out a few native tests, but very quickly we stopped running them. Mainly as we were spending most of our time in a JavaScript environment, so tests would need to run in Xcode. As we don't need a Mac to run our tests, we can use linux CI servers and get 3-4 minute CI runs.
+
+# Deployment
+
+This one is a bit tricky to get your head around at first. As React Native is a client-side library that you don't make source-code changes to, which interacts with JavaScript that you bundle with the app (or use the React Native Packager at dev-time.) Then the JavaScript part of your application is just a file that can be updated, amended or fixed at any time.
+
+This is not dynamically swizzing or fishhooking methods, you can ship new JavaScript code to your app which can interact with the exact same version of your React Native library.
+
+This is what makes it possible to ship bug fixes to your app as fast as the web. We use this, but only for admin users. They can choose the JavaScript for any commit on master, or any active pull request inside a beta version of our app.
```

Comment:
I’d like us to be careful with what we say on this. We don’t ship bug fixes to deployed apps. We do let admins ‘test’ newer builds via this mechanism (and even that I’d phrase in a TestFlight build context).